File: backend/app/schemas/playlist.py
from datetime import datetime

from pydantic import BaseModel, Field


class PlaylistSummaryResponse(BaseModel):
    playlist_id: int
    name: str
    description: str | None
    item_count: int
    created_at: datetime
    updated_at: datetime


class PlaylistListResponse(BaseModel):
    items: list[PlaylistSummaryResponse]
    pagination: "PaginationMeta"


class PlaylistCreateRequest(BaseModel):
    name: str = Field(min_length=1, max_length=500)
    description: str | None = None


class PlaylistUpdateRequest(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=500)
    description: str | None = None


class PlaylistItemResponse(BaseModel):
    playlist_item_id: int
    video_id: int
    title: str
    duration_ms: int
    status: str
    sort_order: int
    has_thumbnail: bool


class PlaylistDetailResponse(BaseModel):
    playlist_id: int
    name: str
    description: str | None
    items: list[PlaylistItemResponse]
    created_at: datetime
    updated_at: datetime


class PlaylistItemInput(BaseModel):
    video_id: int


class PlaylistItemsUpdateRequest(BaseModel):
    items: list[PlaylistItemInput]


class PlaylistPlaybackStartRequest(BaseModel):
    resume: bool = True


class PlaylistPlaybackItemResponse(BaseModel):
    playlist_id: int
    playlist_item_id: int
    video_id: int
    title: str
    duration_ms: int
    mime_type: str
    chunk_count: int
    position_ms: int
    status: str
    sort_order: int
    stream_token: str
    start_chunk: "ChunkMetaResponse"
    has_next: bool
    has_prev: bool


class PlaylistNextItemResponse(BaseModel):
    has_next: bool
    item: PlaylistPlaybackItemResponse | None


class PlaylistPrevItemResponse(BaseModel):
    has_prev: bool
    item: PlaylistPlaybackItemResponse | None


class PaginationMeta(BaseModel):
    page: int
    per_page: int
    total_count: int
    total_pages: int


class ChunkMetaResponse(BaseModel):
    chunk_index: int
    start_time_ms: int
    end_time_ms: int
    byte_length: int


PlaylistListResponse.model_rebuild()
PlaylistPlaybackItemResponse.model_rebuild()

File: frontend/src/types/playlist.ts
export interface PlaylistSummary {
  playlist_id: number
  name: string
  description: string | null
  item_count: number
  created_at: string
  updated_at: string
}

export interface PlaylistItem {
  playlist_item_id: number
  video_id: number
  title: string
  duration_ms: number
  status: string
  sort_order: number
  has_thumbnail: boolean
}

export interface PlaylistDetail {
  playlist_id: number
  name: string
  description: string | null
  items: PlaylistItem[]
  created_at: string
  updated_at: string
}

export interface PlaylistPlaybackItem {
  playlist_id: number
  playlist_item_id: number
  video_id: number
  title: string
  duration_ms: number
  mime_type: string
  chunk_count: number
  position_ms: number
  status: string
  sort_order: number
  stream_token: string
  start_chunk: {
    chunk_index: number
    start_time_ms: number
    end_time_ms: number
    byte_length: number
  }
  has_next: boolean
  has_prev: boolean
}

export interface LastVideoContext {
  video_id: number
  title: string
  duration_ms: number
  position_ms: number
  updated_at: string
}

export interface LastPlaylistContext {
  playlist_id: number
  playlist_name: string
  playlist_item_id: number
  video_id: number
  video_title: string
  duration_ms: number
  position_ms: number
  updated_at: string
}

export interface LastPlayback {
  video: LastVideoContext | null
  playlist: LastPlaylistContext | null
}
